lib/Model/Importer/DebugInfo/PDBImporter.cpp

```diff
@@ -0,0 +1,1093 @@
+  auto InputFile = InputFile::open(Importer.getPDBFile()->getFilePath());
+  if (not InputFile) {
+    revng_log(DILogger, "Unable to open PDB file");
+    return;
+  }
+
+  LinePrinter Printer(/*Indent=*/2, false, nulls(), Filters);
+  const PrintScope HeaderScope(Printer, /*IndentLevel=*/2);
+  PDBSymbolHandler SymbolHandler(Importer, ProcessedTypes, Session, *InputFile);
+  if (iterateSymbolGroups(*InputFile, HeaderScope, SymbolHandler)) {
+    revng_log(DILogger, "Unable to parse symbols.");
+    return;
+  }
+}
+
+void PDBImporterImpl::run(NativeSession &Session) {
+  populateTypes();
+  populateSymbolsWithTypes(Session);
+
+  deduplicateEquivalentTypes(Importer.getModel());
+  promoteOriginalName(Importer.getModel());
+  purgeUnnamedAndUnreachableTypes(Importer.getModel());
+  revng_assert(Importer.getModel()->verify(true));
+}
+
+void PDBImporter::import(const COFFObjectFile &TheBinary) {
+  // Parse debug info and populate types to Model.
+  const codeview::DebugInfo *DebugInfo;
+  StringRef PDBFilePath;
+  std::unique_ptr<IPDBSession> Session;
+  NativeSession *TheNativeSession;
+
+  auto EC = TheBinary.getDebugPDBInfo(DebugInfo, PDBFilePath);
+  if (not EC and DebugInfo != nullptr and not PDBFilePath.empty()) {
+    if (llvm::sys::fs::exists(PDBFilePath)) {
+      auto Err = loadDataForPDB(PDB_ReaderType::Native, PDBFilePath, Session);
+      if (not Err) {
+        TheNativeSession = static_cast<NativeSession *>(Session.get());
+        // TODO: We are using the static_cast due to lack of an LLVM RTTI
+        // support for this. Once it is improved in LLVM, we should avoid this.
+        if (Session->getLoadAddress() != TheNativeSession->getLoadAddress())
+          revng_abort();
+
+        ThePDBFile = &TheNativeSession->getPDBFile();
+      } else {
+        revng_log(DILogger, "Unable to read PDB file.");
+      }
+    } else {
+      // Usualy the PDB files will be generated on a different machine,
+      // so the location read from the debug directory wont be up to date. At
+      // first try to find it in the current `.` dir.
+      // TODO: Introduce a `cl::opt` that will be used for providing a
+      // whitelisted  paths for finding PDB files.
+      auto PDBFileNameOnly = PDBFilePath.substr(PDBFilePath.find_last_of('\\')
+                                                + 1);
+      if (llvm::sys::fs::exists(PDBFileNameOnly.str())) {
+        auto Err = loadDataForPDB(PDB_ReaderType::Native,
+                                  PDBFileNameOnly.str(),
+                                  Session);
+        if (not Err) {
+          TheNativeSession = static_cast<NativeSession *>(Session.get());
+          // TODO: The same as the TODO marker above.
+          if (Session->getLoadAddress() != TheNativeSession->getLoadAddress())
+            revng_abort();
+
+          ThePDBFile = &TheNativeSession->getPDBFile();
+        } else {
+          revng_log(DILogger, "Unable to read PDB file.");
```

Comment:
Duplicated code, please factor it out.

Reply:
I did it in the following commit - `[Model: Add an opt for user defined paths of PDB]`. Can I leave it as is, or rather amend this one?

Reply:
I've squashed it anyway.